neptune/new/internal/backends/hosted_file_operations.py

```diff
@@ -308,64 +310,72 @@ def _multichunk_upload(
-                )
-        return []
-    finally:
-        file_stream.close()
+    while True:
+        try:
+            file_stream = upload_entry.get_stream()
+            entry_length = upload_entry.length()
+            try:
+                if entry_length <= multipart_config.max_single_part_size:
+                    # single upload
+                    data = file_stream.read()
+                    result = upload_raw_data(
+                        http_client=swagger_client.swagger_spec.http_client,
+                        url=urlset.single,
+                        data=data,
+                        query_params=query_params,
+                    )
+                    _attribute_upload_response_handler(result)
+                else:
+                    # chunked upload
+                    result = (
+                        urlset.start_chunked(**query_params, totalLength=entry_length)
+                        .response()
+                        .result
+                    )
+                    if result.errors:
+                        raise MetadataInconsistency(
+                            [err.errorDescription for err in result.errors]
+                        )
+
+                    no_ext_query_params = query_params.copy()
+                    if "ext" in no_ext_query_params:
+                        del no_ext_query_params["ext"]
+
+                    upload_id = result.uploadId
+                    chunker = FileChunker(
+                        upload_entry.source_path,
+                        file_stream,
+                        entry_length,
+                        multipart_config,
+                    )
+                    for idx, chunk in enumerate(chunker.generate()):
+                        result = upload_raw_data(
+                            http_client=swagger_client.swagger_spec.http_client,
+                            url=urlset.send_chunk,
+                            data=chunk.data,
+                            headers={"X-Range": _build_x_range(chunk, entry_length)},
+                            query_params={
+                                "uploadId": upload_id,
+                                "uploadPartIdx": idx,
+                                **no_ext_query_params,
+                            },
+                        )
+                        _attribute_upload_response_handler(result)
+
+                    result = (
+                        urlset.finish_chunked(**no_ext_query_params, uploadId=upload_id)
+                        .response()
+                        .result
+                    )
+                    if result.errors:
+                        raise MetadataInconsistency(
+                            [err.errorDescription for err in result.errors]
+                        )
+                return []
+            finally:
+                file_stream.close()
+        except UploadedFileChanged as e:
```

Comment:
Let's retry in case of 500 http code as well.

Reply:
That doesn't sound good, tbh.

Reply:
1. Why not? It can save Runs from crashing in case of random server errors. How about retrying specific http call instead of whole upload?
2. Can we abort old pending multipart upload in case of error (retrying whole upload) to save space?

Reply:
Ad 2. sure we could, but server has currently not API for that. I'd prefer to classify this as "a future improvement" rather than "part of this fix".
Ad 1. I cringe when I hear "random server errors". There should be no such thing, and it certainly shouldn't be routinely handled in client-side code.

Reply:
- We already have had this discussion few months ago and we decided to retry 500 (this place in code was just missed back then). And after the time I can say it was probably good decision as we saved probably thousends of Runs from crash in case of temporary errors not handled properly in servers. It is very important thing for user experience. We should do what we can to save Runs from crashing.
- We should totally retry at least 502 and 503. Do we do now?
- What are actual cons of retrying 500 besides of you don't liking it? I can't see any if they are not too often.
- Do you actually believe it is possible to have bug-free system of this size?

Reply:
You're right, we missed those API calls in our generic error handling, so this is not the place for this discussion. Fixed, but nevertheless:

> Do you actually believe it is possible to have bug-free system of this size?

I do not. But I strongly believe in failing early instead of going forward no matter what, which can cause invalid state, corrupted data etc. Bugs should be fixed, not hidden.

> What are actual cons of retrying 500 besides of you don't liking it?

500 can and often does mean that server encountered a situation or data batch so unexpected that it crashed. Retrying such request doesn't help, but creates an instant DOS script -- it just retries the request continuously, potentially crashing server instances faster that they recover.

Reply:
> I do not. But I strongly believe in failing early instead of going forward no matter what, which can cause invalid state, corrupted data etc. Bugs should be fixed, not hidden.
They are not hidden. They all are reported.

> I strongly believe in failing early instead of going forward no matter what, which can cause invalid state
We do on our side. But it is not necessary to bother our users.

> 500 can and often does mean that server encountered a situation or data batch so unexpected that it crashed. Retrying such request doesn't help, but creates an instant DOS script -- it just retries the request continuously, potentially crashing server instances faster that they recover.
It is not true if retries are rare.